### PR TITLE
[Rule Tuning] Replace KQL exceptions for Query DSL Exceptions

### LINKED ISSUE #3496: [Rule Tuning] Review KQL Syntax used to exclude Windows Paths
## Description

Related to #3486 

We've been using some syntax to exclude windows paths (like `?\:\\\\`) to avoid wildcards, but it they are causing false positives as this syntax doesn't work in every field. We must review and tune the rules appropriately to solve this.

Works:

310538511-c023ae19-e2ec-45f4-b183-47901ffade5a

Doesn't Work:

image

Works:

image

### PR DESCRIPTION
**Pull request by w0rk3r**
## Issues

Resolves #3496 

## Summary

In the 8.12 release, the syntax we were using to exclude some Windows paths in KQL stopped working. While we didn't investigate further which change caused this, we could determine that by asking the affected users and testing with different versions.

This PR proposes using the Filters functionality to execute [query DSL](https://www.elastic.co/guide/en/elasticsearch/reference/8.11/query-dsl.html) queries to effectively exclude the paths, more specifically using the [wildcard query](https://www.elastic.co/guide/en/elasticsearch/reference/8.11/query-dsl-wildcard-query.html).

This feature is not used broadly in our ruleset, but it is already in production in the Threat Intel Indicator match rules:

* https://github.com/elastic/detection-rules/blob/main/rules/cross-platform/threat_intel_indicator_match_address.toml
* https://github.com/elastic/detection-rules/blob/main/rules/cross-platform/threat_intel_indicator_match_hash.toml
* https://github.com/elastic/detection-rules/blob/main/rules/cross-platform/threat_intel_indicator_match_registry.toml
* https://github.com/elastic/detection-rules/blob/main/rules/cross-platform/threat_intel_indicator_match_url.toml

The main benefits of using this instead of using the new KQL query syntax:

* Support to case-insensitive queries! No need to duplicate values to match different cases, like `Windows` vs `windows` vs `WINDOWS`
* We can use `?` to match random single chars instead of using `*`, which is more expensive.
* No need to use expensive `*` to match white space chars, nor `?`, we can just use the space as the value is quoted.

For the PowerShell rules, we can't switch to other languages as neither ES|QL or EQL support text fields, and we also don't have similar functionality in those languages for new_terms.

## Screenshots

image

image

### CODE REVIEW

File: detection_rules/rule.py
Comment (Mikaayenson) [inline detection_rules/rule.py:586]: Was this feature only added in 8.12? Just trying to understand the minstack.

Comment (Mikaayenson) [inline detection_rules/rule.py:586]: Based on your examples, I also think we need to flush out the schema more with a dataclass.

Comment (w0rk3r) [reply detection_rules/rule.py:586]: This is to not affect older rule versions where the syntax works fine

Comment (terrancedejesus) [reply detection_rules/rule.py:586]: I believe what Mika is discussing here is when was filters first available to use with rules. This way we can min-stack the compatibility within the code for backporting purposes.

Comment (w0rk3r) [reply detection_rules/rule.py:586]: Oh, assuming that threat intel rules use it, and they were created 3 years ago, I suppose it was added in the early 7.x

https://github.com/elastic/detection-rules/pull/1133

**Comment by terrancedejesus** [comment]
@Mikaayenson - This PR is in draft and on hold without the `filters` key being added to the rule schema. Should we follow up and sync on this or is the pipeline a bit full?

**Review by Samirbous — APPROVED** [review]

File: detection_rules/rule.py
Comment (Mikaayenson) [inline detection_rules/rule.py:631]:
```suggestion
```

File: rules/windows/collection_mailbox_export_winlog.toml
Comment (w0rk3r) [inline rules/windows/collection_mailbox_export_winlog.toml:80]:
```
{"wildcard":{"file.path":{"case_insensitive":true,"value":"?:\\\\Users\\\\*\\\\AppData\\\\Roaming\\\\Microsoft\\\\Exchange\\\\RemotePowerShell\\\\*"}}}
```

image

File: rules/windows/collection_mailbox_export_winlog.toml
Comment (w0rk3r) [inline rules/windows/collection_mailbox_export_winlog.toml:87]:
```
{"wildcard":{"file.path":{"case_insensitive":true,"value":"?:\\\\Users\\\\*\\\\AppData\\\\Local\\\\Temp\\\\tmp_????????.???\\\\tmp_????????.???.ps?1"}}}
```
image

File: rules/windows/collection_mailbox_export_winlog.toml
Comment (w0rk3r) [inline rules/windows/collection_mailbox_export_winlog.toml:94]:
```
{"wildcard":{"file.path":{"case_insensitive":true,"value":"?:\\\\Windows\\\\TEMP\\\\tmp_????????.???\\\\tmp_????????.???.ps?1"}}}
```
image

File: rules/windows/collection_posh_clipboard_capture.toml
Comment (w0rk3r) [inline rules/windows/collection_posh_clipboard_capture.toml:113]:
```
{"wildcard":{"file.path":{"case_insensitive":true,"value":"?:\\\\program?files\\\\powershell\\\\?\\\\Modules\\\\*.psd1"}}}
```
image

File: rules/windows/collection_posh_clipboard_capture.toml
Comment (w0rk3r) [inline rules/windows/collection_posh_clipboard_capture.toml:120]:
```
{"wildcard":{"file.path":{"case_insensitive":true,"value":"?:\\\\Windows\\\\system32\\\\WindowsPowerShell\\\\v1.0\\\\Modules\\\\*.psd1"}}}
```

image

File: rules/windows/collection_posh_clipboard_capture.toml
Comment (w0rk3r) [inline rules/windows/collection_posh_clipboard_capture.toml:127]:
```
{"wildcard":{"file.path":{"case_insensitive":true,"value":"?:\\\\Program?Files\\\\WindowsPowerShell\\\\Modules\\\\*.ps?1"}}}
```

image

File: rules/windows/defense_evasion_posh_assembly_load.toml
Comment (w0rk3r) [inline rules/windows/defense_evasion_posh_assembly_load.toml:153]:
```
{"wildcard":{"file.path":{"case_insensitive":true,"value":"C:\\\\Program Files\\\\Microsoft Monitoring Agent\\\\Agent\\\\Health Service State\\\\Monitoring Host Temporary Files*\\\\AvailabilityGroupMonitoring.ps1"}}}
```
image

File: rules/windows/defense_evasion_posh_compressed.toml
Comment (w0rk3r) [inline rules/windows/defense_evasion_posh_compressed.toml:150]:
```
{"wildcard":{"file.path":{"case_insensitive":true,"value":"?:\\\\ProgramData\\\\Microsoft\\\\Windows Defender Advanced Threat Protection\\\\Downloads\\\\*"}}}
```
image

File: rules/windows/discovery_privileged_localgroup_membership.toml
Comment (w0rk3r) [inline rules/windows/discovery_privileged_localgroup_membership.toml:176]:
```
{"wildcard":{"winlog.event_data.CallerProcessName":{"case_insensitive":true,"value":"C:\\\\Program Files (x86)\\\\*.exe"}}}
```
image

File: rules/windows/discovery_privileged_localgroup_membership.toml
Comment (w0rk3r) [inline rules/windows/discovery_privileged_localgroup_membership.toml:183]:
```
{"wildcard":{"winlog.event_data.CallerProcessName":{"case_insensitive":true,"value":"C:\\\\Program Files\\\\*.exe"}}}
```
image

**Review by w0rk3r — COMMENTED** [review]
Adding screenshots of the queries (after exporting the rule) matching the intended values. These queries can be tested on telem cluster, in discover > Add a filter > Edit as Query DSL > Paste the query

image

**Review by terrancedejesus — APPROVED** [review]
LGTM. I'd wait on merging in, pending final review and approval from Mika.

**Review by Aegrah — APPROVED** [review]
LGTM from the RAD perspective.

**Comment by w0rk3r** [comment]
Testing the mechanism by:

Including:

image
image

Excluding:

image
image

**Comment by w0rk3r** [comment]
Exported NDJSON: [rules_export.txt](https://github.com/elastic/detection-rules/files/14774347/rules_export.txt)
Resulting TOML file (Imported from NDJSON):
[exchange_mailbox_export_via_powershell.txt](https://github.com/elastic/detection-rules/files/14774367/exchange_mailbox_export_via_powershell.txt)

**Comment by Mikaayenson** [comment]
@w0rk3r decided to remove the $state in https://github.com/elastic/detection-rules/pull/3505/commits/549254714356f43275f049383e7d6498c24cb228 
- Without significant overhaul to how we manage fields with marshmallow and dataclasses the field renaming will not work for loading/dumping `state` to `$state`. I planned to use the marshmallow pre/post_load/dump methods but things get really ad-hoc trying to go that route. 
- When looking at the field is exported from Kibana, but it can not be configured from the UI

**Review by Mikaayenson — APPROVED** [review]
Based on the exploration and additional manual testing from @w0rk3r, this LGTM. I added some additional thoughts here https://github.com/elastic/detection-rules/pull/3505#issuecomment-2024134605 but I think these additional field schemas are a step in the right direction compared to leaving the schema wide open as it previously was. I also made notes in the schema for places we can expand the schema if needed.

**Review by Mikaayenson — COMMENTED** [review]
@w0rk3r I noticed we do some transformations already within the codebase. I readded the state field and transform at the end when exporting rules. 

@eric-forte-elastic @terrancedejesus would be good to put extra 👀 on this.

File: detection_rules/rule.py
Comment (terrancedejesus) [inline detection_rules/rule.py:279]: I see what you did here for importing for exporting. Makes sense :)

File: detection_rules/rule.py
Comment (terrancedejesus) [inline detection_rules/rule.py:680]: Any issue with importing here if we export from Kibana? I assume the transform method is only called if available from to_api_format if I recall? The schema's themselves should be resilient to this?

Comment (eric-forte-elastic) [reply detection_rules/rule.py:680]: Just to note, I think you are right @terrancedejesus, that it only gets called in `to_api_format` used in `_post_dict_conversion()` [here](https://github.com/elastic/detection-rules/blob/218c3bead6da1463b7a037f7a0ccf8d4df9b669a/detection_rules/rule.py#L997)

```python
        # rule type transforms
        self.data.transform(obj) if hasattr(self.data, 'transform') else False

```

**Review by eric-forte-elastic — APPROVED** [review]
Manual and peer review LGTM :+1:

File: detection_rules/rule.py
Comment (brokensound77) [inline detection_rules/rule.py:276]:
```suggestion
    state: Optional[FilterStateStore] = = field(metadata=dict(data_key="$state"))
```

File: detection_rules/rule.py
Comment (brokensound77) [reply detection_rules/rule.py:279]: You should be able to get simplify and get rid of these 2 methods by using

```
= field(metadata=dict(data_key="$state"))
```

https://github.com/elastic/detection-rules/pull/3505/files#r1546589383 (should be tested)

File: detection_rules/rule.py
Comment (brokensound77) [inline detection_rules/rule.py:737]:
```suggestion
```

File: detection_rules/schemas/definitions.py
Comment (brokensound77) [inline detection_rules/schemas/definitions.py:161]: could change this to a literal to simplify further

`StoreType = Literal['appState', 'globalState']`

File: detection_rules/rule.py
Comment (Mikaayenson) [reply detection_rules/rule.py:276]: Did you get that to actually work? I tried that initially and it didn't work for me.

Comment (brokensound77) [reply detection_rules/rule.py]: as long as it is being loaded via marshmallow it should work (in our case, it is the from_dict within our mixins)

Comment (Mikaayenson) [reply detection_rules/rule.py]: 🤷 I think with the other three bugs, I must have mixed up the branch when originally testing. I just retested and everything looks good export/import.

Comment (Mikaayenson) [reply detection_rules/rule.py]: [rules_export-7.ndjson.txt](https://github.com/elastic/detection-rules/files/14827017/rules_export-7.ndjson.txt)

Note to test. I had to use one of the [patches](https://github.com/elastic/detection-rules/issues/3542#issuecomment-2024302296) that addresses a new_terms import bug.

**Comment by w0rk3r** [comment]
Last test before merge:

Exported from TOML: [20240401T173236L.txt](https://github.com/elastic/detection-rules/files/14828317/20240401T173236L.txt)
Exported from Kibana and imported to TOML: [exchange_mailbox_export_via_powershell.txt](https://github.com/elastic/detection-rules/files/14828310/exchange_mailbox_export_via_powershell.txt)

**Pull request merged**
